### PR TITLE
Add replicated tensor type and handle conv2d with sharded input channels

### PR DESCRIPTION
**Pull request by sogartar**

### CODE REVIEW

File: sharktank/tests/ops/sharded_test.py [hunk 1]
```diff
@@ -13,6 +13,74 @@
...
+        ]
+        expected_result = torch.cat(shards, dim=shard_dim)
+
+        sharded = ShardedPrimitiveTensor(shard_dim=shard_dim, ts=shards)
+        actual_result = ops.all_gather(sharded)
+
+        for shard in actual_result.shards:
+            torch.testing.assert_close(shard.as_torch(), expected_result)
+
+    def testCov2dShardedInputAndOutputChannelsOneGroup(self):
```
Comment (stellaraccident) [inline sharktank/tests/ops/sharded_test.py:31]: Nit: misspelled "testCov"

**Review by stellaraccident — APPROVED** [review]
Nice, thanks!

**Comment by stellaraccident** [comment]
Discussed offline about possibly changing `__eq__` to something like `is_deep_equivalent` and documenting specifically what that means.

**Comment by sogartar** [comment]
@stellaraccident I renamed `__eq__` to `is_deep_equal`.

**Comment by sogartar** [comment]
I rebased and resolved conflicts.

**Comment by sogartar** [comment]
@stellaraccident, I ran into this problem that when loading the dataset if we mem map, then even when deleting the Python objects the underlying files are still open. I see that the files are memory mapped in IREE. I assume we are missing somewhere an ownership relation.

**Pull request merged**